core/shared/src/main/scala/fs2/concurrent/Channel.scala

```diff
@@ -116,174 +118,179 @@ sealed trait Channel[F[_], A] {
+
+  // used as a marker to wake up q.take when the channel is closed
+  private[this] val Sentinel = new AnyRef
+
+  private[this] val LeftClosed: Either[Channel.Closed, Unit] = Left(Channel.Closed)
+  private[this] val RightUnit: Either[Channel.Closed, Unit] = Right(())
+
+  // technically this should be A | Sentinel.type
+  // the queue will consist of exclusively As until we shut down, when there will be one Sentinel
+  private[this] def impl[F[_]: Concurrent, A](q: Queue[F, AnyRef]): F[Channel[F, A]] =
+    (Concurrent[F].ref(0), Concurrent[F].ref(false), Concurrent[F].deferred[Unit]).mapN {
+      (leasesR, closedR, closedLatch) =>
+        new Channel[F, A] {
+
+          private[this] val LeftClosedF = LeftClosed.pure[F]
+          private[this] val FalseF = false.pure[F]
+
+          // might be interesting to try to optimize this more, but it needs support from CE
+          val sendAll: Pipe[F, A, Nothing] =
+            _.evalMapChunk(send(_))
+              .takeWhile(_.isRight)
+              .onComplete(Stream.exec(close.void))
+              .drain
+
+          // setting the flag means we won't accept any more sends
+          val close: F[Either[Channel.Closed, Unit]] =
+            closedR.getAndSet(true).flatMap { b =>
+              if (b) {
+                LeftClosedF
+              } else {
+                leasesR.get.flatMap { leases =>
+                  if (leases <= 0)
+                    q.offer(Sentinel).start.as(RightUnit)
                   else
-                    (
-                      State(values, size, None, (a, producer) :: producers, false),
-                      notifyStream(waiting).as(rightUnit) <* waitOnBound(producer, poll)
-                    )
-              }.flatten
+                    RightUnit.pure[F]
+                }
+              }
             }
-          }
 
-        def trySend(a: A) =
-          state.modify {
-            case s @ State(_, _, _, _, closed @ true) =>
-              (s, Channel.closed[Boolean].pure[F])
-
-            case s @ State(values, size, waiting, producers, closed @ false) =>
-              if (size < capacity)
-                (
-                  State(a :: values, size + 1, None, producers, false),
-                  notifyStream(waiting).as(rightTrue)
-                )
-              else
-                (s, rightFalse.pure[F])
-          }.flatten
-
-        def close =
-          state
-            .modify {
-              case s @ State(_, _, _, _, closed @ true) =>
-                (s, Channel.closed[Unit].pure[F])
-
-              case State(values, size, waiting, producers, closed @ false) =>
-                (
-                  State(values, size, None, producers, true),
-                  notifyStream(waiting).as(rightUnit) <* signalClosure
-                )
+          val isClosed: F[Boolean] = closedR.get
+
+          // there are four states to worry about: open, closing, draining, quiesced
```

Comment:
State machine makes sense to me. It's a bit implicit in the various `F[Boolean]`, but reifying it a bit more depends on the decision on splitting Refs

Reply:
Also the fact that `Queue` is separate. We can't really reify the state machine, but we can probably make the documentation more clear, and also perhaps by pulling things out into better-named functions to indicate the state transitions and tests.